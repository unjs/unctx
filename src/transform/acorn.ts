import * as acorn from "acorn";
import MagicString from "magic-string";
import { walk } from "estree-walker";
import type {
  Node,
  CallExpression,
  BlockStatement,
  AwaitExpression,
  Position,
} from "estree";
import {
  type TransformerOptions,
  defaultTransformerOptions,
  kInjected,
} from "./index.js";

type MaybeHandledNode = Node & {
  [kInjected]?: boolean;
};

export function createTransformer(options: TransformerOptions = {}) {
  options = {
    ...defaultTransformerOptions,
    ...options,
  };

  const objectDefinitionFunctions = Object.keys(options.objectDefinitions!);

  const matchRE = new RegExp(
    `\\b(${[...options.asyncFunctions!, ...objectDefinitionFunctions].join(
      "|",
    )})\\(`,
  );

  function shouldTransform(code: string) {
    return typeof code === "string" && matchRE.test(code);
  }

  function transform(code: string, options_: { force?: false } = {}) {
    if (!options_.force && !shouldTransform(code)) {
      return;
    }
    const ast = acorn.parse(code, {
      sourceType: "module",
      ecmaVersion: "latest",
      locations: true,
    });

    const s = new MagicString(code);
    const lines = code.split("\n");

    let detected = false;

    walk(ast as any, {
      enter(node: Node) {
        if (node.type === "CallExpression") {
          const functionName = _getFunctionName(node.callee);
          if (options.asyncFunctions!.includes(functionName)) {
            transformFunctionArguments(node);
            if (functionName !== "callAsync") {
              const lastArgument = node.arguments[node.arguments.length - 1];
              if (lastArgument && lastArgument.loc) {
                s.appendRight(toIndex(lastArgument.loc.end), ",1");
              }
            }
          }
          if (objectDefinitionFunctions.includes(functionName)) {
            for (const argument of node.arguments) {
              if (argument.type !== "ObjectExpression") {
                continue;
              }

              for (const property of argument.properties) {
                if (
                  property.type !== "Property" ||
                  property.key.type !== "Identifier"
                ) {
                  continue;
                }

                if (
                  options.objectDefinitions![functionName]?.includes(
                    property.key?.name,
                  )
                ) {
                  transformFunctionBody(property.value);
                }
              }
            }
          }
        }
      },
    });

    if (!detected) {
      return;
    }

    s.appendLeft(
      0,
      `import { ${options.helperName} as __executeAsync } from "${options.helperModule}";`,
    );

    return {
      code: s.toString(),
      magicString: s,
    };

    function toIndex(pos: Position) {
      return lines.slice(0, pos.line - 1).join("\n").length + pos.column + 1;
    }

    function transformFunctionBody(function_: Node) {
      if (
        function_.type !== "ArrowFunctionExpression" &&
        function_.type !== "FunctionExpression"
      ) {
        return;
      }

      // No need to transform non-async function
      if (!function_.async) {
        return;
      }

      const body = function_.body as BlockStatement;

      let injectVariable = false;
      walk(body, {
        enter(
          node: MaybeHandledNode,
          parent: MaybeHandledNode | undefined | null,
        ) {
          if (node.type === "AwaitExpression" && !node[kInjected]) {
            detected = true;
            injectVariable = true;
            injectForNode(node, parent);
          } else if (
            node.type === "IfStatement" &&
            node.consequent.type === "ExpressionStatement" &&
            node.consequent.expression.type === "AwaitExpression"
          ) {
            detected = true;
            injectVariable = true;
            (node.consequent.expression as MaybeHandledNode)[kInjected] = true;
            injectForNode(node.consequent.expression, node);
          }
          // Skip transform for nested functions
          if (
            node.type === "ArrowFunctionExpression" ||
            node.type === "FunctionExpression" ||
            node.type === "FunctionDeclaration"
          ) {
            return this.skip();
          }
        },
      });

      if (injectVariable && body.loc) {
        s.appendLeft(toIndex(body.loc.start) + 1, "let __temp, __restore;");
      }
    }

    function transformFunctionArguments(node: CallExpression) {
      for (const function_ of node.arguments) {
        transformFunctionBody(function_);
      }
    }

    function injectForNode(
      node: AwaitExpression,
      parent: Node | undefined | null,
    ) {
      const isStatement = parent?.type === "ExpressionStatement";

      if (!node.loc || !node.argument.loc) {
        return;
      }

      s.remove(toIndex(node.loc.start), toIndex(node.argument.loc.start));
      s.remove(toIndex(node.loc.end), toIndex(node.argument.loc.end));

      s.appendLeft(
        toIndex(node.argument.loc.start),
        isStatement
          ? `;(([__temp,__restore]=__executeAsync(()=>`
          : `(([__temp,__restore]=__executeAsync(()=>`,
      );
      s.appendRight(
        toIndex(node.argument.loc.end),
        isStatement
          ? `)),await __temp,__restore());`
          : `)),__temp=await __temp,__restore(),__temp)`,
      );
    }
  }

  return {
    transform,
    shouldTransform,
  };
}

function _getFunctionName(node: Node): string {
  if (node.type === "Identifier") {
    return node.name;
  } else if (node.type === "MemberExpression") {
    return _getFunctionName(node.property);
  }
  return "";
}
