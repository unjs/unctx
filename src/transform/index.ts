export interface TransformerOptions {
  /**
   * The function names to be transformed.
   *
   * @default ['withAsyncContext']
   */
  asyncFunctions?: string[];
  /**
   * @default 'unctx'
   */
  helperModule?: string;
  /**
   * @default 'executeAsync'
   */
  helperName?: string;
  /**
   * Whether to transform properties of an object defined with a helper function. For example,
   * to transform key `middleware` within the object defined with function `defineMeta`, you would pass:
   * `{ defineMeta: ['middleware'] }`.
   * @default {}
   */
  objectDefinitions?: Record<string, string[]>;
}

export const kInjected = "__unctx_injected__";

export type MaybeHandledNode = Node & {
  [kInjected]?: boolean;
};

export const defaultTransformerOptions: TransformerOptions = {
  asyncFunctions: ["withAsyncContext"],
  helperModule: "unctx",
  helperName: "executeAsync",
  objectDefinitions: {},
};
