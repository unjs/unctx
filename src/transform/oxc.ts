/**
 * Based on transform.ts but uses oxc-parser and oxc-walker
 */
import { parseSync } from "oxc-parser";
import MagicString from "magic-string";
import { walk } from "oxc-walker";
import type {
  Node,
  CallExpression,
  BlockStatement,
  AwaitExpression,
} from "oxc-parser";
import {
  defaultTransformerOptions,
  kInjected,
  type TransformerOptions,
} from "./index.js";

type MaybeHandledNode = Node & {
  [kInjected]?: boolean;
};

export function createTransformer(options: TransformerOptions = {}) {
  options = {
    ...defaultTransformerOptions,
    ...options,
  };

  const objectDefinitionFunctions = Object.keys(options.objectDefinitions!);

  const matchRE = new RegExp(
    `\\b(${[...options.asyncFunctions!, ...objectDefinitionFunctions].join(
      "|",
    )})\\(`,
  );

  function shouldTransform(code: string) {
    return typeof code === "string" && matchRE.test(code);
  }

  function transform(code: string, options_: { force?: false } = {}) {
    if (!options_.force && !shouldTransform(code)) {
      return;
    }
    const ast = parseSync("", code, {
      sourceType: "module",
    });

    const s = new MagicString(code);

    let detected = false;

    walk(ast.program, {
      enter(node: Node) {
        if (node.type === "CallExpression") {
          const functionName = _getFunctionName(node.callee);
          if (options.asyncFunctions!.includes(functionName)) {
            transformFunctionArguments(node);
            if (functionName !== "callAsync") {
              const lastArgument = node.arguments[node.arguments.length - 1];
              if (lastArgument && lastArgument.end) {
                s.appendRight(lastArgument.end, ",1");
              }
            }
          }
          if (objectDefinitionFunctions.includes(functionName)) {
            for (const argument of node.arguments) {
              if (argument.type !== "ObjectExpression") {
                continue;
              }

              for (const property of argument.properties) {
                if (
                  property.type !== "Property" ||
                  property.key.type !== "Identifier"
                ) {
                  continue;
                }

                if (
                  options.objectDefinitions![functionName]?.includes(
                    property.key?.name,
                  )
                ) {
                  transformFunctionBody(property.value);
                }
              }
            }
          }
        }
      },
    });

    if (!detected) {
      return;
    }

    s.appendLeft(
      0,
      `import { ${options.helperName} as __executeAsync } from "${options.helperModule}";`,
    );

    return {
      code: s.toString(),
      magicString: s,
    };

    function transformFunctionBody(function_: Node) {
      if (
        function_.type !== "ArrowFunctionExpression" &&
        function_.type !== "FunctionExpression"
      ) {
        return;
      }

      // No need to transform non-async function
      if (!function_.async) {
        return;
      }

      const body = function_.body as BlockStatement;

      let injectVariable = false;
      walk(body, {
        enter(
          node: MaybeHandledNode,
          parent: MaybeHandledNode | undefined | null,
        ) {
          if (node.type === "AwaitExpression" && !node[kInjected]) {
            detected = true;
            injectVariable = true;
            injectForNode(node, parent);
          } else if (
            node.type === "IfStatement" &&
            node.consequent.type === "ExpressionStatement" &&
            node.consequent.expression.type === "AwaitExpression"
          ) {
            detected = true;
            injectVariable = true;
            (node.consequent.expression as MaybeHandledNode)[kInjected] = true;
            injectForNode(node.consequent.expression, node);
          }
          // Skip transform for nested functions
          if (
            node.type === "ArrowFunctionExpression" ||
            node.type === "FunctionExpression" ||
            node.type === "FunctionDeclaration"
          ) {
            return this.skip();
          }
        },
      });

      if (injectVariable && body.start) {
        s.appendLeft(body.start + 1, "let __temp, __restore;");
      }
    }

    function transformFunctionArguments(node: CallExpression) {
      for (const function_ of node.arguments) {
        transformFunctionBody(function_);
      }
    }

    function injectForNode(
      node: AwaitExpression,
      parent: Node | undefined | null,
    ) {
      const isStatement = parent?.type === "ExpressionStatement";

      if (!node.start || !node.argument.start) {
        return;
      }

      s.remove(node.start, node.argument.start);
      s.remove(node.end, node.argument.end);

      s.appendLeft(
        node.argument.start,
        isStatement
          ? `;(([__temp,__restore]=__executeAsync(()=>`
          : `(([__temp,__restore]=__executeAsync(()=>`,
      );
      s.appendRight(
        node.argument.end,
        isStatement
          ? `)),await __temp,__restore());`
          : `)),__temp=await __temp,__restore(),__temp)`,
      );
    }
  }

  return {
    transform,
    shouldTransform,
  };
}

function _getFunctionName(node: Node): string {
  if (node.type === "Identifier") {
    return node.name;
  } else if (node.type === "MemberExpression") {
    return _getFunctionName(node.property);
  }
  return "";
}
