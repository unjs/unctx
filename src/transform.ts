
import * as acorn from 'acorn'
import MagicString from 'magic-string'
import { walk } from 'estree-walker'
import type {
  Node, CallExpression,
  BlockStatement,
  AwaitExpression,
  Position
} from 'estree'

export interface TransformerOptions {
  /**
   * The function names to be transformed.
   *
   * @default ['withAsyncContext', 'callAsync']
   */
  asyncFunctions?: string[]
  /**
   * @default 'unctx'
   */
  helperModule?: string
  /**
   * @default 'executeAsync'
   */
  helperName?: string
}

export function createTransformer (options: TransformerOptions = {}) {
  options = {
    asyncFunctions: ['withAsyncContext', 'callAsync'].concat(options.asyncFunctions || []),
    helperModule: 'unctx',
    helperName: 'executeAsync',
    ...options
  }

  const matchRE = new RegExp(`\\b(${options.asyncFunctions.join('|')})\\(`)

  function shouldTransform (code: string) {
    return typeof code === 'string' && matchRE.test(code)
  }

  function transform (code: string, opts: { force?: false } = {}) {
    if (!opts.force && !shouldTransform(code)) {
      return
    }
    const ast = acorn.parse(code, {
      sourceType: 'module',
      ecmaVersion: 'latest',
      locations: true
    })

    const s = new MagicString(code)
    const lines = code.split('\n')

    let detected = false

    walk(ast, {
      enter (node: Node) {
        if (node.type === 'CallExpression') {
          const functionName = getFunctionName(node.callee)
          if (options.asyncFunctions.includes(functionName)) {
            transformFunctionBody(node)
            if (functionName !== 'callAsync') {
              const lastArg = node.arguments[node.arguments.length - 1]
              if (lastArg) {
                s.appendRight(toIndex(lastArg.loc.end), ',1')
              }
            }
          }
        }
      }
    })

    if (!detected) {
      return null
    }

    s.appendLeft(0, `import { ${options.helperName} as __executeAsync } from "${options.helperModule}";`)

    return {
      code: s.toString(),
      magicString: s
    }

    function getFunctionName (node: Node) {
      if (node.type === 'Identifier') {
        return node.name
      } else if (node.type === 'MemberExpression') {
        return getFunctionName(node.property)
      }
    }

    function toIndex (pos: Position) {
      return lines.slice(0, pos.line - 1).join('\n').length + pos.column + 1
    }

    function transformFunctionBody (node: CallExpression) {
      for (const fn of node.arguments) {
        if (fn.type !== 'ArrowFunctionExpression' && fn.type !== 'FunctionExpression') {
          continue
        }

        // No need to transform non-async function
        if (!fn.async) {
          continue
        }

        const body = fn.body as BlockStatement

        let injectVariable = false
        walk(body, {
          enter (node: Node, parent: Node | undefined) {
            if (node.type === 'AwaitExpression') {
              detected = true
              injectVariable = true
              injectForNode(node, parent)
            }
            // Skip transform for nested functions
            if (node.type === 'ArrowFunctionExpression' || node.type === 'FunctionExpression' || node.type === 'FunctionDeclaration') {
              return this.skip()
            }
          }
        })

        if (injectVariable) {
          s.appendLeft(
            toIndex(body.loc.start) + 1,
            'let __temp, __restore;'
          )
        }
      }
    }

    function injectForNode (node: AwaitExpression, parent: Node | undefined) {
      const body = code.slice(
        toIndex(node.argument.loc.start),
        toIndex(node.argument.loc.end)
      )

      const isStatement = parent?.type === 'ExpressionStatement'

      s.overwrite(
        toIndex(node.loc.start),
        toIndex(node.loc.end),
        isStatement
          ? `;(([__temp,__restore]=__executeAsync(()=>${body})),await __temp,__restore());`
          : `(([__temp,__restore]=__executeAsync(()=>${body})),__temp=await __temp,__restore(),__temp)`
      )
    }
  }

  return {
    transform,
    shouldTransform
  }
}
