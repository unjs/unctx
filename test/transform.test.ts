import { expect, it, describe } from "vitest";
import { createTransformer } from "../src/transform/acorn";

describe("transforms", () => {
  const transformer = createTransformer({
    asyncFunctions: ["withAsyncContext", "callAsync"],
    objectDefinitions: {
      defineSomething: ["someKey"],
    },
  });

  function transform(input: string) {
    return transformer.transform(
      // Slice 6 spaces indention for snapshot alignment
      input
        .split("\n")
        .map((index) => index.slice(6))
        .join("\n"),
    )?.code;
  }

  it("transforms", () => {
    expect(
      transform(`
      export default withAsyncContext(async () => {
        const ctx1 = useSomething()
        await something()
        const ctx2 = useSomething()
      })
    `),
    ).toMatchSnapshot();
  });

  it("transforms await as variable", () => {
    expect(
      transform(`
      export default withAsyncContext(async () => {
        const foo = await something()
        const bar = hello(await something())
        const ctx = useSomething()
      })
    `),
    ).toMatchSnapshot();
  });

  it("transforms await in nested scopes", () => {
    expect(
      transform(`
      export default withAsyncContext(async () => {
        for (const i of foo) {
          if (i) {
            await i()
          }
        }
        const ctx = useSomething()
      })
    `),
    ).toMatchSnapshot();
  });

  it("transforms await in try-catch", () => {
    expect(
      transform(`
      export default withAsyncContext(async () => {
        let user;

        try {
          user = await fetchUser();
        } catch (e) {
          user = null;
        }

        if (!user)
          return navigateTo('/');
      })
    `),
    ).toMatchSnapshot();
  });

  it("transforms dot usage", () => {
    expect(
      transform(`
      export default ctx.callAsync(async () => {
        const ctx1 = useSomething()
        await something()
        const ctx2 = useSomething()
      })
    `),
    ).toMatchSnapshot();

    expect(
      transform(`
      export default x.ctx.callAsync(async () => {
        const ctx1 = useSomething()
        await something()
        const ctx2 = useSomething()
      })
    `),
    ).toMatchSnapshot();
  });

  it("does not transform non async usage", () => {
    expect(
      transform(`
      export default withAsyncContext(async () => {
        const ctx = useSomething()
      })
    `),
    ).toBeUndefined();
  });

  it("does not transform unrelated nested functions", () => {
    expect(
      transform(`
      export default withAsyncContext(async () => {
        async function foo() {
          await something()
        }
        const bar = async () => {
          await something()
        }
        const ctx = useSomething()
      })
    `),
    ).toBeUndefined();
  });

  it("transforms validly nested functions", () => {
    expect(
      transform(`
      export default withAsyncContext(async () => {
        await something()

        withAsyncContext(async () => {
          await something()
        })

        const ctx = useSomething()
      })
    `),
    ).toMatchSnapshot();
  });

  it("transforms multiple awaits in same chunk", () => {
    expect(
      transform(`
      export default withAsyncContext(async () => {
        await writeConfig(await readConfig())
      })
    `),
    ).toMatchSnapshot();
  });

  it("does not transform non target function", () => {
    expect(
      transform(`
      export default someFunction(async () => {
        const ctx1 = useSomething()
        await something()
        const ctx2 = useSomething()
      })
    `),
    ).toBeUndefined();
  });

  it("transforms certain keys of an object", () => {
    expect(
      transform(`
      export default defineSomething({
        someKey: async () => {
          const ctx1 = useSomething()
          await something()
          const ctx2 = useSomething()
        },
        async someKey () {
          const ctx1 = useSomething()
          await something()
          const ctx2 = useSomething()
        },
        ...someKey,
        someKey: 421,
        someKey () {
          const ctx1 = useSomething()
          const ctx2 = useSomething()
        },
        async someOtherKey () {
          const ctx1 = useSomething()
          await something()
          const ctx2 = useSomething()
        }
      })
    `),
    ).toMatchSnapshot();
  });

  it("doesn't transform non-objects", () => {
    expect(
      transform(`
      export default defineSomething('test')
    `),
    ).toBeUndefined();
  });
  it("Should not add a statement terminator if expression comes after if statement", () => {
    expect(
      transform(`
      export default withAsyncContext(async () => {
        if(false) await something()
      })
    `),
    ).toMatchSnapshot();
  });
});
