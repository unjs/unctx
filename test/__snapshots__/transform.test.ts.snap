// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transforms > Should not add a statement terminator if expression comes after if statement 1`] = `
"import { executeAsync as __executeAsync } from "unctx";
export default withAsyncContext(async () => {let __temp, __restore;
  if(false) (([__temp,__restore]=__executeAsync(()=>something())),__temp=await __temp,__restore(),__temp)
},1)
"
`;

exports[`transforms > transforms 1`] = `
"import { executeAsync as __executeAsync } from "unctx";
export default withAsyncContext(async () => {let __temp, __restore;
  const ctx1 = useSomething()
  ;(([__temp,__restore]=__executeAsync(()=>something())),await __temp,__restore());
  const ctx2 = useSomething()
},1)
"
`;

exports[`transforms > transforms await as variable 1`] = `
"import { executeAsync as __executeAsync } from "unctx";
export default withAsyncContext(async () => {let __temp, __restore;
  const foo = (([__temp,__restore]=__executeAsync(()=>something())),__temp=await __temp,__restore(),__temp)
  const bar = hello((([__temp,__restore]=__executeAsync(()=>something())),__temp=await __temp,__restore(),__temp))
  const ctx = useSomething()
},1)
"
`;

exports[`transforms > transforms await in nested scopes 1`] = `
"import { executeAsync as __executeAsync } from "unctx";
export default withAsyncContext(async () => {let __temp, __restore;
  for (const i of foo) {
    if (i) {
      ;(([__temp,__restore]=__executeAsync(()=>i())),await __temp,__restore());
    }
  }
  const ctx = useSomething()
},1)
"
`;

exports[`transforms > transforms await in try-catch 1`] = `
"import { executeAsync as __executeAsync } from "unctx";
export default withAsyncContext(async () => {let __temp, __restore;
  let user;

  try {
    user = (([__temp,__restore]=__executeAsync(()=>fetchUser())),__temp=await __temp,__restore(),__temp);
  } catch (e) {
    user = null;
  }

  if (!user)
    return navigateTo('/');
},1)
"
`;

exports[`transforms > transforms certain keys of an object 1`] = `
"import { executeAsync as __executeAsync } from "unctx";
export default defineSomething({
  someKey: async () => {let __temp, __restore;
    const ctx1 = useSomething()
    ;(([__temp,__restore]=__executeAsync(()=>something())),await __temp,__restore());
    const ctx2 = useSomething()
  },
  async someKey () {let __temp, __restore;
    const ctx1 = useSomething()
    ;(([__temp,__restore]=__executeAsync(()=>something())),await __temp,__restore());
    const ctx2 = useSomething()
  },
  ...someKey,
  someKey: 421,
  someKey () {
    const ctx1 = useSomething()
    const ctx2 = useSomething()
  },
  async someOtherKey () {
    const ctx1 = useSomething()
    await something()
    const ctx2 = useSomething()
  }
})
"
`;

exports[`transforms > transforms dot usage 1`] = `
"import { executeAsync as __executeAsync } from "unctx";
export default ctx.callAsync(async () => {let __temp, __restore;
  const ctx1 = useSomething()
  ;(([__temp,__restore]=__executeAsync(()=>something())),await __temp,__restore());
  const ctx2 = useSomething()
})
"
`;

exports[`transforms > transforms dot usage 2`] = `
"import { executeAsync as __executeAsync } from "unctx";
export default x.ctx.callAsync(async () => {let __temp, __restore;
  const ctx1 = useSomething()
  ;(([__temp,__restore]=__executeAsync(()=>something())),await __temp,__restore());
  const ctx2 = useSomething()
})
"
`;

exports[`transforms > transforms multiple awaits in same chunk 1`] = `
"import { executeAsync as __executeAsync } from "unctx";
export default withAsyncContext(async () => {let __temp, __restore;
  ;(([__temp,__restore]=__executeAsync(()=>writeConfig((([__temp,__restore]=__executeAsync(()=>readConfig())),__temp=await __temp,__restore(),__temp)))),await __temp,__restore());
},1)
"
`;

exports[`transforms > transforms validly nested functions 1`] = `
"import { executeAsync as __executeAsync } from "unctx";
export default withAsyncContext(async () => {let __temp, __restore;
  ;(([__temp,__restore]=__executeAsync(()=>something())),await __temp,__restore());

  withAsyncContext(async () => {let __temp, __restore;
    ;(([__temp,__restore]=__executeAsync(()=>something())),await __temp,__restore());
  },1)

  const ctx = useSomething()
},1)
"
`;
